### PR TITLE
RuleFactory: accept strings for time-based rules

### PR DESCRIPTION
**Pull request by amy**
This is just a first idea of what should happen. I have some notes in the code marked as Code Review for input. I also need some date / time samples to test. 

EDIT: ignore this comment

### CODE REVIEW

File: src/RuleFactory.php [hunk 1]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
 
   } // _getRuleParam
 
+  /**
+   * @param mixed $dateTime
+   * @return \DateTimeImmutable
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: empty line before @return

File: src/RuleFactory.php [hunk 2]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+   */
+  private static function _getDateImmutableObject( $dateTime ) {
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: variables should use `$under_score`, or simply lowercase all together `$datetime`

File: src/RuleFactory.php [hunk 3]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+
+    if ( is_a( $dateTime, '\DateTimeImmutable' ) ) {
+
+      return $dateTime;
+
+    } // if is \DateTimeImmutable
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: no need for comment on this small scope

File: src/RuleFactory.php [hunk 4]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
 
   } // _getRuleParam
 
+  /**
+   * @param mixed $dateTime
+   * @return \DateTimeImmutable
+   */
+  private static function _getDateImmutableObject( $dateTime ) {
+
+    if ( is_a( $dateTime, '\DateTimeImmutable' ) ) {
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: is_a is slow, you can use `instanceof`

http://stackoverflow.com/questions/3017684/what-is-the-difference-between-is-a-and-instanceof

File: src/RuleFactory.php [hunk 5]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+
+      return new \DateTimeImmutable( "$dateTime" );
+
+    } // if is string
+
+    if ( is_int( $dateTime ) ) {
+
+      return new \DateTimeImmutable( "@$dateTime" );
+
+    } // if Unix epoch time int
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: what happens if it get's down here and it doesn't match any of the conditionals?

File: src/RuleFactory.php [hunk 6]
```diff
@@ -82,4 +91,36 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+
+    } // if is \DateTimeImmutable
+
+    if ( is_string( $dateTime ) ) {
+
+      if ( is_numeric( $dateTime ) ) {
+
+        return new \DateTimeImmutable( "@$dateTime" );
+
+      } // if Unix epoch time string
```
Comment (ingluisjimenez) [inline src/RuleFactory.php]: no need for comments

File: tests/unit/RuleFactoryTest.php [hunk 7]
```diff
@@ -90,33 +90,129 @@ public function createBetweenTimeRuleSuccess() {
...
+    $this->assertInstanceOf( BetweenTimesRule::class, $rule );
+
+  } // createBetweenTimeRuleStringSuccess
+
+  /**
+   * @test
+   */
+  public function createBetweenTimeRuleStringFormatFail() {
+
+    $this->expectException( \Exception::class );
```
Comment (ingluisjimenez) [inline tests/unit/RuleFactoryTest.php]: this should be a custom exception on our code, if it doesn't fall into any of the conditionals

File: tests/unit/RuleFactoryTest.php [hunk 8]
```diff
@@ -129,18 +225,59 @@ public function createStartTimeRuleFail() {
 
   } // createStartTimeRuleFail
 
+  public function createStartTimeRuleStringFormatFail() {
+
+    $this->expectException( \Exception::class );
```
Comment (ingluisjimenez) [inline tests/unit/RuleFactoryTest.php]: same as above

File: tests/unit/RuleFactoryTest.php [hunk 9]
```diff
@@ -153,6 +290,16 @@ public function createEndTimeRuleFail() {
 
   } // createEndTimeRuleFail
 
+  public function createEndTimeRuleStringFormatFail() {
+
+    $this->expectException( \Exception::class );
```
Comment (ingluisjimenez) [inline tests/unit/RuleFactoryTest.php]: same as above, custom exception

File: src/RuleFactory.php [hunk 10]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
 
+  /**
+   * @param mixed $date_time
+   *
+   * @return \DateTimeImmutable
+   */
+  private static function _getDateImmutableObject( $date_time ) {
+
+    if ( $date_time instanceof \DateTimeImmutable ) {
+
```
Comment (be-dmitry) [inline src/RuleFactory.php]: remove line break

File: src/RuleFactory.php [hunk 11]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+      return $date_time;
+
```
Comment (be-dmitry) [inline src/RuleFactory.php]: remove line break

File: src/RuleFactory.php [hunk 12]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+      } // if date time is string
+
+      if ( is_int( $date_time ) ) {
+
+        return new \DateTimeImmutable( "@$date_time" );
+
+      }
+    } // try date time format
+
+    // @codingStandardsIgnoreStart
```
Comment (be-dmitry) [inline src/RuleFactory.php]: Is there really nothing more specific that you can catch here? We should avoid doing this if at all possible

File: src/RuleFactory.php [hunk 13]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+
+      return $date_time;
+
+    }
+
+    try {
+      if ( is_string( $date_time ) ) {
+
+        if ( is_numeric( $date_time ) ) {
+
```
Comment (be-dmitry) [inline src/RuleFactory.php]: remove line break

File: src/RuleFactory.php [hunk 14]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+          return new \DateTimeImmutable( "@$date_time" );
+
```
Comment (be-dmitry) [inline src/RuleFactory.php]: remove line break

File: src/RuleFactory.php (hunk 12 above)
Comment (amy) [inline src/RuleFactory.php]: There's nothing more specific. =(

File: src/RuleFactory.php [hunk 15]
```diff
@@ -82,4 +92,48 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+        }
+
+        return new \DateTimeImmutable( "$date_time" );
+
+      } // if date time is string
+
+      if ( is_int( $date_time ) ) {
```
Comment (DFoxinator) [inline src/RuleFactory.php]: I thin the is_numeric call and is_int call can be combined by using ctype_digit instead: http://php.net/manual/en/function.ctype-digit.php ctype_digit((string)$date_time)

Comment (be-dmitry) [inline src/RuleFactory.php]: +1

**Comment by DFoxinator** [comment]
👍

File: src/RuleFactory.php [hunk 16]
```diff
@@ -82,4 +92,38 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+    } // try date time format
+
+    // @codingStandardsIgnoreStart
+    catch( \Exception $e ) {
+    // @codingStandardsIgnoreEnd
+
+      throw new DateTimeImmutableException( "Invalid date_time format. $date_time given." );
+
+    } // catch date time excpetion
+
```
Comment (be-dmitry) [inline src/RuleFactory.php]: Sorry for nitpicking, but remove this extra line break

File: src/RuleFactory.php [hunk 17]
```diff
@@ -82,4 +92,38 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+        return new \DateTimeImmutable( "$date_time" );
+      }
+
+    } // try date time format
+
+    // @codingStandardsIgnoreStart
+    catch( \Exception $e ) {
+    // @codingStandardsIgnoreEnd
+
+      throw new DateTimeImmutableException( "Invalid date_time format. $date_time given." );
```
Comment (be-dmitry) [inline src/RuleFactory.php]: Wrap `$date_time` in curly bvraces as such `{$date_time}`

File: src/RuleFactory.php [hunk 18]
```diff
@@ -82,4 +92,38 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
 
   } // _getRuleParam
 
+  /**
+   * @param mixed $date_time
+   *
```
Comment (be-dmitry) [inline src/RuleFactory.php]: Need `@throws` docblock line

File: tests/unit/RuleFactoryTest.php [hunk 19]
```diff
@@ -129,18 +226,59 @@ public function createStartTimeRuleFail() {
...
 
     $this->assertInstanceOf( EndTimeRule::class, $rule );
 
-  } // createEndTimeRuleSuccess
+  } // createEndTimeRuleObjectSuccess
+
+  public function createEndTimeRuleEpochStringSuccess() {
+
+    $rule = RuleFactory::create( EndTimeRule::RULE_NAME, [
+        'end' => "1465238714"
```
Comment (DFoxinator) [inline tests/unit/RuleFactoryTest.php]: just a nitpick, but for strings with no vars we should remain consistent and use single quotes (seems like you used them in most places)

File: tests/unit/RuleFactoryTest.php [hunk 20]
```diff
@@ -90,33 +91,129 @@ public function createBetweenTimeRuleSuccess() {
...
-  } // createStartTimeRuleSuccess
+  } // createStartTimeRuleObjectSuccess
+
+  /**
+   * @test
+   */
+  public function createStartTimeRuleEpochStringSuccess() {
+
+    $rule = RuleFactory::create( StartTimeRule::RULE_NAME, [
+        'start' => "1465238714"
```
Comment (DFoxinator) [inline tests/unit/RuleFactoryTest.php]: single quotes

File: tests/unit/RuleFactoryTest.php [hunk 21]
```diff
@@ -90,33 +91,129 @@ public function createBetweenTimeRuleSuccess() {
...
+
+
+  /**
+   * @test
+   */
+  public function createBetweenTimeRuleEpochStringSuccess() {
+
+    $rule = RuleFactory::create( BetweenTimesRule::RULE_NAME, [
+        'from' => "1465238714",
+        'to'   => "1465238715",
```
Comment (DFoxinator) [inline tests/unit/RuleFactoryTest.php]: nit pick - single quotes

File: src/RuleFactory.php [hunk 22]
```diff
@@ -82,4 +92,38 @@ private static function _getRuleParam( $param_name, $rule_type, array $params )
...
+        return new \DateTimeImmutable( "{$date_time}" );
+      }
+
+    } // try date time format
+
+    // @codingStandardsIgnoreStart
+    catch( \Exception $e ) {
+    // @codingStandardsIgnoreEnd
+      throw new DateTimeImmutableException( "Invalid date_time format. {$date_time} given." );
+    } // catch date time excpetion
```
Comment (be-dmitry) [inline src/RuleFactory.php]: No need for this comment if block is < 3 lines

Comment (ingluisjimenez) [inline src/RuleFactory.php]: DateTimeImmutableException sounds too specific IMO, its just a datetime, immutable is an implementation detail, and thinking more abstract, it is just an invalid parameter exception.

**Comment by be-dmitry** [comment]
👍

File: src/RuleFactory.php (hunk 22 above)
Comment (amy) [inline src/RuleFactory.php]: the linter + test suite will yell at me if I dont. =(

**Pull request merged**